spark/src/main/scala/geotrellis/spark/io/tms/TMSReader.scala

```diff
@@ -0,0 +1,84 @@
+package geotrellis.spark.io.tms
+
+import geotrellis.raster._
+import geotrellis.raster.render._
+import geotrellis.spark.SpatialKey
+import geotrellis.vector.{Point, Extent}
+import geotrellis.spark.tiling.ZoomedLayoutScheme
+
+import spire.syntax.cfor._
+import scalaj.http._
+
+import scala.collection.mutable._
+import java.net.{URI, URL}
+import javax.imageio._
+
+
+class TMSReader[T](
+  uriTemplate: String,
+  f: URI => T = TMSReader.decodeTile _
+) extends Serializable {
+  def uri(z: Int, x: Int, y: Int): URI = new URI(
+    uriTemplate.replace("{z}", z.toString)
+      .replace("{x}", x.toString)
+      .replace("{y}", y.toString)
+  )
+
+  def read(zoom: Int, bounds: GridBounds): Iterator[(SpatialKey, T)] = {
+    // construct bounds from zoom + extent
+    for ((x, y) <- bounds.coordsIter)
+    yield SpatialKey(x, y) -> f(uri(zoom, x, y))
+  }
+
+  def read(zoom: Int, extent: Extent, tileSize: Int = ZoomedLayoutScheme.DEFAULT_TILE_SIZE): Iterator[(SpatialKey, T)] = {
+    val layoutDefinition = ZoomedLayoutScheme.layoutForZoom(zoom, extent, tileSize)
+    val bounds = layoutDefinition.mapTransform.extentToBounds(extent)
+    read(zoom, bounds)
+  }
+
+  def read(zoom: Int, x: Int, y: Int): T = f(uri(zoom, x, y))
+}
+
+object TMSReader {
+
+  def apply[T](uriTemplate: String, f: URI => T = TMSReader.decodeTile _): TMSReader[T] =
+    new TMSReader(uriTemplate, f)
+
+  // ImageIO use here to deal with different decoding behaviors (png vs jpg)
+  def decodeTile(uri: URI): Tile = {
+    val img = ImageIO.read(uri.toURL)
+    val height = img.getHeight()
+    val width = img.getWidth()
+    val rgbArr = img.getRGB(0, 0, width, height, null, 0, width)
+    IntArrayTile(rgbArr, width, height, None)
+  }
+
+  // ImageIO use here to deal with different decoding behaviors (png vs jpg)
+  def decodeMultibandTile(uri: URI): MultibandTile = {
+    val img = ImageIO.read(uri.toURL)
+    val height = img.getHeight()
+    val width = img.getWidth()
+    val length = width * height
+
+    val rgbArr = img.getRGB(0, 0, width, height, null, 0, width)
+    val reds = new ArrayBuffer[Byte](length)
+    val greens = new ArrayBuffer[Byte](length)
+    val blues = new ArrayBuffer[Byte](length)
+    val alphas = new ArrayBuffer[Byte](length)
+    cfor(0)(_ < length, _ + 1) { cellIdx =>
+      val cell = rgbArr(cellIdx)
+      reds :+ cell.red.toByte
+      greens :+ cell.green.toByte
+      blues :+ cell.blue.toByte
+      alphas :+ cell.alpha.toByte
+    }
+
+    val redTile = ByteArrayTile(reds.toArray, width, height, None)
+    val greenTile = ByteArrayTile(greens.toArray, width, height, None)
+    val blueTile = ByteArrayTile(blues.toArray, width, height, None)
+    val alphaTile = ByteArrayTile(alphas.toArray, width, height, None)
```

Comment:
Hm, there is this conversion, it's an extra `copyToArray` (actually here it would be just a while loop to iterate over all elements of an `ArrayBuffer`), mb just to use `Array` instead of `ArrayBuffer`?

Reply:
I also suggest rewriting this just to use `Array[Byte]`, e.g.:

```scala
val reds = Array.ofDim[Byte](length)
... 
cfor(0)(_ < length, _ + 1) { cellIdx =>
   val cell = rgbArr(cellIdx)
   reds(cellIdx) = cell.red.toByte
...
}
...
val redTile = ArrayTile(reds, width, height)
```